newsfragments/2512.breaking.rst

```diff
@@ -0,0 +1,18 @@
+:func:`trio.move_on_after` and :func:`trio.fail_after` now sets the deadline upon entering the context manager, instead of at initialization. The vast majority of usage initialized them when entering and is therefore unaffected. But if you have code that looks like:
+
+.. code-block:: python3
+
+    cs = trio._move_on_after(5)
+    ...
+    with cs:
+        ...
```

Comment:
Couldn't we detect (most of) this usage? Might help this work better. IE compare construction time vs usage time? (and then a kwarg to silence + enable this new behavior?)

I'd just like something a bit more backwards compatible or at least that doesn't fail silently if people used it wrong.

Reply:
Yeah that's probably a good call :+1:

Reply:
Adding the functionality to `move_on_after` was quite straightforward.... but `fail_after` is being a pita, since `@contextmanager` makes the function not get executed at all until we try to enter it. The only way I could figure out how to resolve it was to change `fail_after` to be a full-on class-based contextmanager.